### DISCUSSION TITLE
How to get Instant time deltas in queries?

### DISCUSSION THREAD

**Discussion by MichaelHorn-BuyersEdge** [ravendb/ravendb#13209] [Q&A]
```
{
   "WorkflowStage": "stage1",
   "StageChangeTime": "2021-06-14T15:37:42.6333058Z"
},
{
   "WorkflowStage": "stage2",
   "StageChangeTime": "2021-06-13T15:37:42.6333058Z"
},
```
```
 public String WorkflowStage { get; private set; }
 public Instant? StageChangeTime { get; internal set; }
```

etc

I am looking for a way to calculate the average stage age given this schema.
The problem is that stageTimeChange is in instant and Sum(Instant) doesn't work since it isnt an integer.

       var instantNow = SystemClock.Instance.GetCurrentInstant();
       var result = await Query<InvoiceWorkflow>()
                    .GroupBy(x => new
                    {
                        x.WorkflowStage,
                        x.StageChangeTime

                    })
                    .Select(x => new
                    {
                        count = x.Count(),
                        stage = x.Key.WorkflowStage,
                        sum = x.Sum(i => (instantNow - (i.StageChangeTime == null ? instantNow : i.StageChangeTime.Value)).Seconds)
                    })
                    .ToListAsync();

**Returns values, but the sums are always zero.** 

```
using System.Linq;
using NodaTime;
using Raven.Client.Documents.Indexes;

namespace Domain.InvoiceWorkflow
{
    public class InvoiceWorkflowQueueAgeIndex : AbstractIndexCreationTask<InvoiceWorkflow, InvoiceWorkflowQueueAgeGrouping>
    {
        public override IndexDefinition CreateIndexDefinition()
        {
            var def = base.CreateIndexDefinition();
            def.OutputReduceToCollection = "InvoiceWorkflowQueueAgeGrouping";
            return def;
        }

        public InvoiceWorkflowQueueAgeIndex()
        {
            var nowInstant = SystemClock.Instance.GetCurrentInstant();


            Map = invoiceWorkflows => from invoiceWorkflow in invoiceWorkflows
                select new
                {
                    WorkflowStage = invoiceWorkflow.WorkflowStage,
                    AveAge = ( nowInstant - ( invoiceWorkflow.StageChangeTime.HasValue?invoiceWorkflow.StageChangeTime.Value: nowInstant)).Seconds,
                    Count = 1
                };

            Reduce = invoiceWorkflows 
                => from item in invoiceWorkflows
                    group item by item.WorkflowStage
                into g
                select new
                {
                    WorkflowStage = g.Key,
                    AveAge = g.Sum(x => x.DeltaTime) / g.Sum(x => 1),
                    Count = g.Sum(x => 1)
                };
        }

    }

    public class InvoiceWorkflowQueueAgeGrouping
    {
        public string WorkflowStage { get; set; }
        public int DeltaTime { get; set; }
        public int Count { get; set; }
    }
}
```

**Throws: Failed to compile index name = InvoiceWorkflowQueueAgeIndex**

**Comment by ayende** [comment]
What is the actual query that you are trying to answer? You want the average age? But from what point in time?
Also, that isn't actually something that you need to do at the query level, is it? You are basically just wanting a projection. 

In your queries, you'll always get the separate value for each stage and time, so I assume that there is no actual group by?

**Reply by MichaelHorn-BuyersEdge** [reply]
The specific query is to return the average age of the "StageChangeElapsedTime" which would be defined by "Today - StageChangeTime" and the WorkFlowStage. I was trying to do group by WorkflowStage and StageChangeTime, and return the WorkflowStage and average StageChangeElapsedTime. I am a complete novice in RavenDB, so I probably don't have the vernacular down.

I mean just returning a grouping of WorkFlowStage, Count, and ElapsedTime, then doing a Sum of the ElapsedTime might get me what I want, but Sum(Instant) isnt supported.

**Reply by ayende** [reply]
Is `WorkFlowStage` containing multiple values or just one?

The actual issue you have here is the that "now" is different each query.

Instead, you are better off calculating the minimum date for each stage, then computing the different off that.

//map
from iw in docs.InvoiceWorkflows
select new 
{
  iw.StageChangeTime,
  iw.WorkflowStage
}

// reduce
from r in results
group r by r.WorkflowStage into g
select new
{
    WorkflowStage = g.Key,
    StageChangeTime = g.OrderBy(x=>x.StageChangeTime).First()
}
